goose.go

```diff
@@ -1051,136 +1064,142 @@ func (c *cursor) Remainder() []ast.Stmt {
-		retUnit := coq.ReturnExpr{coq.Tt}
-		ife.Then = coq.BlockExpr{[]coq.Binding{
-			coq.NewAnon(ife.Then),
-			coq.NewAnon(retUnit),
-		}}
-		ife.Else = retUnit
-		return coq.NewAnon(ife)
-	}
-	elseEndsWithReturn := endsWithReturn(s.Else)
-	// the if expression is last in the surrounding block,
-	// so returns in it become returns from the overall function
-	// OR
-	// neither branch terminates the outer function,
-	// the if expression is just a conditional in the middle
-	if !remaining || (!bodyEndsWithReturn && !elseEndsWithReturn) {
-		if s.Else == nil {
-			ife.Else = coq.ReturnExpr{coq.Tt}
-			return coq.NewAnon(ife)
-		}
-		elseExpr, ok := ctx.stmt(s.Else, c, loopVar).Unwrap()
-		if !ok {
-			ctx.nope(s.Else, "if statement else ends with an assignment")
-			return coq.Binding{}
+		Cond: condExpr,
+	}
+
+	// Determine if we want to propagate our usage (i.e. the availability of control effects)
+	// to the conditional or not. This is not relevant for correctness (stmt generation will
+	// appropriately error when a control effect arises where it is not supported), but Crucial
+	// for supporting a reasonable subset of Go.
+	bodyEndsWithReturn := ctx.endsWithReturn(s.Body)
+
+	// If the "then" branch want to early-return, translate it accordingly, propagating the usage.
+	if bodyEndsWithReturn {
+		// stmts takes care of finalization for us
+		ife.Then = ctx.blockStmt(s.Body, usage)
+		if len(remainder) > 0 {
+			// If there are remaining expressions after this, put them into "else".
+			// Actual "else" must be empty.
+			if s.Else != nil {
+				ctx.futureWork(s.Else, "early return in if with an else branch")
+				return coq.Binding{}
+			}
+			ife.Else = ctx.stmts(remainder, usage)
+		} else {
+			// Remainder is empty; translate "else" branch with our usage.
+			if s.Else == nil {
+				// Be careful to perform proper finalization according to our usage.
+				ife.Else = ctx.stmts([]ast.Stmt{}, usage)
+			} else {
+				ife.Else = ctx.stmts(s.Else.(*ast.BlockStmt).List, usage)
+			}
 		}
-		ife.Else = elseExpr
 		return coq.NewAnon(ife)
 	}
-	// there are some other cases that can be handled but it's a bit tricky
-	ctx.futureWork(s, "non-terminal if expressions are only partially supported")
-	return coq.Binding{}
+
+	// No early return in "then" branch; translate this as a conditional in the middle of
+	// a block (not propagating the usage). This will implicitly check that
+	// the "else" branch does not rely on control effects.
+	ife.Then = ctx.blockStmt(s.Body, ExprValLocal)
+	if s.Else == nil {
+		ife.Else = ctx.stmts([]ast.Stmt{}, ExprValLocal)
+	} else {
+		ife.Else = ctx.stmts(s.Else.(*ast.BlockStmt).List, ExprValLocal)
+	}
+	// And translate the remainder with our usage.
+	tailExpr := ctx.stmts(remainder, usage)
+	// Prepend the if-then-else before the tail.
+	bindings := append([]coq.Binding{coq.NewAnon(ife)}, tailExpr.Bindings...)
+	return coq.NewAnon(coq.BlockExpr{ Bindings: bindings })
```

Comment:
The conversion between binding and block and so on here is somewhat annoying, I feel I am not using these types entirely as intended...